### PR TITLE
GH-118095: Add tier 2 support for YIELD_VALUE

### PR DESCRIPTION
**Pull request by markshannon**
This PR:
* Makes a few changes to make `YIELD_VALUE` suitable for tier 2
* Adds a few conditions to the tier 2 optimizer pipeline to support the new `_YIELD_VALUE` micro-op.
* Adds a case for `_FOR_ITER_GEN_FRAME` to the tier 2 optimizer to get rid of a compiler warning. This isn't strictly necessary for this PR, but is just good housekeeping.

* Issue: gh-118095

### CODE REVIEW

**Comment by markshannon** [comment]
The [stats](https://github.com/faster-cpython/benchmarking-public/blob/main/results/bm-20240429-3.13.0a6%2B-3e1b870-PYTHON_UOPS/bm-20240429-azure-x86_64-faster%252dcpython-tier_2_yield_value-3.13.0a6%2B-3e1b870-pystats-vs-base.md) show that this is working as intended. `YIELD_VALUE` disappears as an unsupported opcode for tier 2, and the number of trace "underflows" (unmatched return or yields) increases accordingly.

File: Python/bytecodes.c [hunk 1]
```diff
@@ -1089,32 +1089,38 @@ dummy_func(
...
-        tier1 inst(YIELD_VALUE, (retval -- unused)) {
+        inst(YIELD_VALUE, (retval -- value)) {
             // NOTE: It's important that YIELD_VALUE never raises an exception!
             // The compiler treats any exception raised here as a failed close()
             // or throw() call.
+            #if TIER_ONE
             assert(frame != &entry_frame);
-            frame->instr_ptr = next_instr;
+            #endif
+            frame->instr_ptr++;
```
Comment (iritkatriel) [inline Python/bytecodes.c:1099]: Would be good if the DSL had a symbol for "this instruction's size".

**Review by iritkatriel — APPROVED** [review]
I don't understand the implications for tier2, but for tier1 this looks fine.

**Review by Fidget-Spinner — APPROVED** [review]
I don't understand the tier 1 changes that well, but the tier 2 changes LGTM.

File: Python/bytecodes.c (hunk 1 above)
Comment (markshannon) [reply Python/bytecodes.c:1099]: https://github.com/python/cpython/issues/118423

File: Python/optimizer.c [hunk 2]
```diff
@@ -763,7 +763,7 @@ translate_bytecode_to_trace(
                                 Py_FatalError("garbled expansion");
                         }
 
-                        if (uop == _POP_FRAME || uop == _RETURN_GENERATOR) {
+                        if (uop == _POP_FRAME || uop == _RETURN_GENERATOR || uop == _YIELD_VALUE) {
```
Comment (gvanrossum) [inline Python/optimizer.c:766]: @markshannon Shouldn't this same change be made to optimizer_analysis.c (in get_code())?

**Comment by corona10** [comment]
>Adds a case for _FOR_ITER_GEN_FRAME to the tier 2 optimizer to get rid of a compiler warning. This isn't strictly necessary for this PR, but is just good housekeeping.

For your information, from gcc-14, since it enabled `-Werror=incompatible-pointer-types` by default, the code caused the build failure, but thanks to this PR, everything is okay now.

**Pull request merged**